### ISSUE TITLE
Compare runs from different experiments

### ISSUE THREAD

**Issue by johny-c** [mlflow/mlflow#1414]
### Feature request
I would like to be able to compare runs belonging to different experiments. Currently runs from different experiments cannot be selected simultaneously.

**Comment by rquintino** [comment]
much needed, otherwise limits the value we get from mlflow ui (or am I missing something?).
In the same project we can have several different experiments, but really every run should be comparable with any other run within the same backend storage. Up to the team to decide what is comparable, mlflow should not impose same experiment restriction?

Could be rather easy fix? Experiment should work more like a filter, so we would have an "all experiments" option on the left pane, and corresponding experiment id/experiment name filter on main run list.

Meantime, thinking on some workarounds, moving runs temporarily to a single experiment? or drop experiment level altogether and instead use tags for that.
Thanks!

**Comment by Ben-Epstein** [comment]
@rquintino Out of curiosity what are you using Experiments for in your ML organization? You say "In the same project we can have several different experiments"

Why not call an use the Experiment as the project?

**Comment by rquintino** [comment]
hi @Ben-Epstein sure, experiments can be different approaches/hypothesis we're trying on a specific project. Multiple team members, putting all these tries in same experiment would be a hell to manage and explore.

note: each project gets its own mlflow runs folder, much easier , no need to share this data among projects, works quite well actually.

still very frequently, and by these reasons, we need to compare runs from multiple experiments.

**Comment by Ben-Epstein** [comment]
@rquintino that's very interesting. It seems to me, if I'm understanding you correctly, that a concept of "teams" may make sense as a solution to your problem as well.

The idea of experiments being different approaches to the same project confuses me a bit, as I consider runs to be different approaches to a given experiment. Without knowing the details of your team it's hard to fully digest, but I'm sure you have good reason for your approach.

I wonder, if you treated an experiment as a "project", but only saw the runs (attempts) of people in your team, or you could filter out different teams, would that also work?

The reason I'm asking is because I would love something like teams to be implemented in MLFlow, with permissions of views and UI logins for each member. And it's just another conceptual way to potentially organize your work. 

What are your thoughts?

**Comment by jaklan** [comment]
@Ben-Epstein I'll describe my use-case, which is for me a crucial argument for comparing runs between experiments (unless I miss sth):
- we have a multi-step ML pipeline
- the pipeline is run every Sunday
- each step of the pipeline is running as a separate task within Airflow DAG
- we want to have all steps tracked in one "catalog" - either an experiment or a parent run
- this "catalog" should be named with a date, e.g. "21-04-2020"

And now - to get such an effect with experiments I can just add:
```
mlflow.set_experiment(os.getenv("CURRENT_DATE"))
mlflow.start_run(run_name="step-name")
...
mlflow.end_run()
```
to each step and that's all. But then I cannot compare the final model from "21-04-2020" with the one from e.g. "28-04-2020".

And what about a parent run? How to manage a parent run's lifecycle through many independent steps being "child steps"? I have to start it each time before the proper child run? Moreover - do I really need to pass to every child run a parent run ID looking like `399760f34e6145d3af66b8f04d3ea7d1`? I know I can manage it via envar, but it's far, far less elegant than just using string name like with experiments.

And despite the above - it's just much more intuitive to have the each full pipeline as a single "experiment" on the right side, named with a date and showing all steps as runs. Imho nested runs should be a solution for very complex / detailed workflows, not sth you have to use on a daily basis. Experiments are extremely underused currently.

And btw - if I want to delete a "catalog" from one day it's much easier to find a proper experiment ID (just integer) than this awful run ID. And by delete I mean "true delete" from a hard drive, not just changing metadata of an experiment/run via GUI.

**Comment by Ben-Epstein** [comment]
@jaklan that's interesting, I think I understand your process and issue now. 

Is each stage of the pipeline something you can "compare"? Ie does each stage have metrics? For a stage that's not necessarily a model, how do you know if it's outperforming a different run within that stage?

Also, you say you want to have all steps in one "catalog" which is fair, but I'm curious why. Would the organization structure of:
`mlflow.set_experiment(step_name)
mlflow.start_run(run_name=<>)`

Not make sense? The date is already tracked in every run automatically. If you can't compare different stages of the pipeline why have them in the same "experiment?" That way you can use the run_name for something else that may be useful. 

Regarding nested runs, I think that could actually be an elegant solution as well. Each run is a stage in the pipeline, and it has a bunch of children runs. Comparison within runs of different dates would just be 1 click and you can hold everything in 1 experiment.

Also, regarding the searching by run_id, I fully agree it's not a nice way to do things. We wrote a helper function called `get_runs_by_name` which returns the run/runs with a given name, which may help solve that problem.

**Comment by jaklan** [comment]
@Ben-Epstein 

- >Is each stage of the pipeline something you can "compare"?

    Nope
- >Ie does each stage have metrics?

    Some of them

- >For a stage that's not necessarily a model, how do you know if it's outperforming a different run within that stage?

    There's no need to compare if it's not a model - it works as a "trace", with some metadata and interim artefacts we want store for e.g. debugging purposes. 

- > Would the organization structure of:
mlflow.set_experiment(step_name) mlflow.start_run(run_name=<>)
Not make sense?

    That's an option, but it's far more intuitive for us to just have an experiment view where we can have a quick look at the pipeline from start to end (we see if the step finished, we see timestamps etc.) and be able to explore it. Top-down approach.

- > If you can't compare different stages of the pipeline why have them in the same "experiment?" 

    As above - Data Scientists don't have to look at all at Airflow with such approach, we get MLflow benefits and a quick look at pipeline execution in one place. Well, almost get - because of lack of run comparison between experiments ;)

- > Comparison within runs of different dates would just be 1 click and you can hold everything in 1 experiment.

    Yeah, that's an option, the one which sounds most attractive to me with current capabilities, but what about these technical issues I mentioned? How to organise a "long-living" parent run along independent Python scripts?

**Comment by Ben-Epstein** [comment]
@jaklan Thanks for all of the answers! Very interesting, and a good example of where it may very well be useful to have cross experiment run comparisons.  

The helper function is pretty straightforward, but it's in our custom tracking plugin. We essentially loop through all runs of a requested experiment (or all runs of all experiments if no experiment is passed in) and check the run name of each run. If it matches, return the run id/run ids. Not efficient but for this specific task efficiency isn't really a concern.

**Comment by pvillacorta-stratio** [comment]
@jaklan +1 , I am doing exactly the same. One experiment per day, one run per stage. From one day to another, the source data may slightly change, or the way it is processed / joined may change. We are logging parameters concerning this pre-processing + join stage. Cleaning data and fitting models would be subsequent stages. It would be very nice to compare the same stage across different days (experiments)

**Comment by CuriousSingularity** [comment]
@jaklan + 1, this is a nice feature to have, as the comparison between the different runs from different experiment is essential.

**Comment by Dudly01** [comment]
+1 to the feature. Currently I have to resort to the workaround of _copying runs into a dummy experiment and modifying the meta.yaml files_ to get something similar.

**Comment by zabzug-pfpt** [comment]
+1 here! Same usecase of wanting to use the same "baseline" run(s) across multiple experiments.

**Comment by mohammedayub44** [comment]
+1 for inter-experiment comparison.

**Comment by WillJStone** [comment]
+1 for inter-experiment comparison

**Comment by ZeyuSun** [comment]
+1 for inter-experiment comparison

**Comment by mk-gojek** [comment]
+1 for inter-experiment comparison

**Comment by notTheRealSanta** [comment]
+1 for inter-experiment comparison.

**Comment by julled** [comment]
+1 for inter-experiment comparison.

**Comment by BenWilson2** [comment]
With naught but a 3 year delay, this is being implemented in #5971
